File: src/cache.rs
use std::cmp::min;
use std::vec;

use crate::cache_client::{CacheClient, CacheResult};
use crate::ann_index::ANNIndex;
use crate::types::EmbeddingData;
use crate::utils::get_unix_seconds;

pub const EMBEDDING_DIMENSION: usize = 1536;

pub struct Cache<'a> {
    client: Box<dyn CacheClient>,
    ann_index: Box<dyn ANNIndex<'a>>,
    embedding_size: usize,
    index_initialized: bool,
    current_id: usize,
    cache_ttl: i64,
    redis_key: String,
}

impl <'a> Cache<'a> {
    pub fn new(client: Box<dyn CacheClient>, ann_index: Box<dyn ANNIndex<'a>>, redis_key: String, embedding_size: usize, cache_ttl: i64) -> Self {
        Self {
            client: client,
            ann_index: ann_index,
            embedding_size: embedding_size,
            index_initialized: false,
            current_id: 0,
            cache_ttl: cache_ttl,
            redis_key: redis_key,
        }
    }


    fn load_index(&mut self) -> CacheResult<()> {
        let data = self.get_all_embeddings()?;
        
        if data.len() == 0 {
            self.ann_index.init_index(1, EMBEDDING_DIMENSION)?;
            self.index_initialized = true;
            return Ok(());
        }
        
        if (data[0].embedding.len() == 0) {
            return Err("Something went wrong!".into());
        }
        
        self.ann_index.init_index(data.len(), EMBEDDING_DIMENSION)?;
        
        let max_val = data.iter().map(|d|(d.id)).max().unwrap();
        for d in data {
            self.ann_index.add_pt(d.embedding, d.id)?;
        }

        self.current_id = max_val + 1;
        self.index_initialized = true;

        Ok(())
    }


    pub fn store_embedding(&mut self, query: String, embedding: Vec<f32>, response: String) -> CacheResult<()>{
        if embedding.len() != self.embedding_size {
            return Err("embedding sizes mismatch!".into());
        }

        println!("Storing embedding");

        let new_embedding = embedding.clone();

        let id = self.current_id;

        let timestamp = get_unix_seconds();

        let data = EmbeddingData{id, query, embedding, response, timestamp };

        self.client.h_set(&self.redis_key, &id.to_string(), &serde_json::to_string(&data)?)?;
        
        if self.cache_ttl != 0 {
            self.client.expire(&self.redis_key,self.cache_ttl)?;
        }

        if (!self.index_initialized) {
            println!("Initializing index");
            self.ann_index.init_index(1000, EMBEDDING_DIMENSION)?;
            self.index_initialized = true;
        }

        let current_elements = self.ann_index.get_curr_ct()?;
        let max_elements = self.ann_index.get_max_elements()?;

        if (current_elements > max_elements) {
            println!("Resizing index");
            self.ann_index.resize(current_elements + 1000)?;
        }
        
        println!("Adding point to index");
        self.ann_index.add_pt(new_embedding, id)?;

        Ok(())
    }

    pub fn semantic_search(&mut self, embedding: &[f32], k: usize) -> CacheResult<Vec<EmbeddingData>> {
        if embedding.len() != self.embedding_size {
            return Err("embedding sizes mismatch!".into());
        }

        if !self.index_initialized {
            self.load_index()?;
        }

        let current_elements = self.ann_index.get_curr_ct()?;
        let adjusted_k = min(k, current_elements);

        if adjusted_k == 0 {
            return Ok(vec![]);
        }

        let result = self.ann_index.search_knn(embedding, adjusted_k)?;
        let ids: Vec<String> = result.iter().map(|n| n.0.to_string()).collect();        
        let id_refs: Vec<&str> = ids.iter().map(|s| s.as_str()).collect();
        let data: Vec<Option<String>> = self.client.hm_get(&self.redis_key, &id_refs)?;

        let mut parsed_data = Vec::new();
        for maybe_json in data {
            if let Some(json_str) = maybe_json {
                let embedding_data: EmbeddingData = serde_json::from_str(&json_str)?;
                parsed_data.push(embedding_data);
            }
        }

        return Ok(parsed_data);
    }

    pub fn get_all_embeddings(&mut self) -> CacheResult<Vec<EmbeddingData>> {
        let data = self.client.h_get_all(&self.redis_key)?;

        let embeddings = data.into_iter()
            .map(|(_, v)| serde_json::from_str(&v))
            .filter_map(Result::ok)
            .collect();

        // Do not need to remap embeddings back into numbers?
        
        return Ok(embeddings);
    }

}

fn mapper((_, v): (String, String)) -> CacheResult<EmbeddingData> {
    let embedding_data: EmbeddingData = serde_json::from_str(&v)?;
    return Ok(embedding_data);
}

File: src/cache.py
from typing import List
from .cache_client import CacheClient
from .ann_index import ANNIndex
from .custom_types import EmbeddingData
from .utils import get_unix_seconds, logger

from dataclasses import dataclass
from typing import Dict, List
from .cache_client import CacheClient
from .ann_index import ANNIndex
from .custom_types import EmbeddingData
from .utils import get_unix_seconds, logger

@dataclass
class CacheConfig:
    client: CacheClient
    ann_index: ANNIndex
    embedding_size: int
    current_id: int = 0
    index_initialized: bool = False
    initial_size: int = 1000

class EmbeddingCache:
    def __init__(
        self,
        configs: Dict[str, CacheConfig],
        redis_key_prefix: str,
        cache_ttl: int,
    ):
        """
        configs: map from modality name (e.g. "text", "image") to its CacheConfig
        redis_key_prefix: base key (we’ll append the modality)
        cache_ttl: seconds until each hash key expires
        """
        self.configs = configs
        self.redis_key_prefix = redis_key_prefix
        self.cache_ttl = cache_ttl


    def _redis_key(self, modality: str) -> str:
        return f"{self.redis_key_prefix}:{modality}"


    def _load_index(self, modality: str) -> None:
        cfg = self.configs[modality]
        logger.debug(f"Loading '{modality}' index")
        data = self.get_all_embeddings(modality)
        if not data and not cfg.index_initialized:
            cfg.ann_index.init_index(cfg.initial_size, cfg.embedding_size)
        else:
            cfg.ann_index.init_index(len(data), cfg.embedding_size)
            max_id = 0
            for d in data:
                cfg.ann_index.add_pt(d.embedding, d.id)
                max_id = max(max_id, d.id)
            cfg.current_id = max_id + 1
        cfg.index_initialized = True


    def store_embedding(
        self,
        modality: str,
        query: str,
        embedding: List[float],
        response: str,
    ) -> None:
        cfg = self.configs[modality]
        if len(embedding) != cfg.embedding_size:
            logger.error(f"Size mismatch for {modality}: {len(embedding)} vs {cfg.embedding_size}")
            raise ValueError("Embedding size mismatch")

        eid = cfg.current_id
        cfg.current_id += 1

        payload = EmbeddingData(
            id=eid,
            query=query,
            embedding=embedding,
            response=response,
            timestamp=get_unix_seconds()
        )

        payload_json = payload.model_dump_json()

        key = self._redis_key(modality)
        cfg.client.h_set(key, str(eid), payload_json)
        if self.cache_ttl:
            cfg.client.expire(key, self.cache_ttl)

        if not cfg.index_initialized:
            cfg.ann_index.init_index(cfg.initial_size, cfg.embedding_size)
            cfg.index_initialized = True

        if cfg.ann_index.get_curr_ct() >= cfg.ann_index.get_max_elements():
            new_size = cfg.ann_index.get_curr_ct() + 1000
            logger.debug(f"Resizing '{modality}' ANN index → {new_size}")
            cfg.ann_index.resize(new_size)

        cfg.ann_index.add_pt(embedding, eid)

    def semantic_search(
        self,
        modality: str,
        embedding: List[float],
        k: int
    ) -> List[EmbeddingData]:
        cfg = self.configs[modality]
        if len(embedding) != cfg.embedding_size:
            logger.error(f"Size mismatch for {modality}: {len(embedding)} vs {cfg.embedding_size}")
            raise ValueError("Embedding size mismatch")

        if not cfg.index_initialized:
            self._load_index(modality)

        curr = cfg.ann_index.get_curr_ct()
        if curr == 0:
            return []

        k = min(k, curr)
        results = cfg.ann_index.search_knn(embedding, k)
        ids = [str(r[0]) for r in results]
        raw = cfg.client.hm_get(self._redis_key(modality), ids)

        return [
            EmbeddingData.parse_raw(item)
            for item in raw
            if item is not None
        ]


    def get_all_embeddings(self, modality: str) -> List[EmbeddingData]:
        raw = self.configs[modality].client.h_get_all(self._redis_key(modality))
        return [EmbeddingData.parse_raw(v) for v in raw.values()]
